### PR TITLE
Support data parallelism (static)

### PR DESCRIPTION
**Pull request by Ying1123**
Add built-in data parallelism support. It can give better throughput then tensor parallelism when there is enough memory.

## Usage
Usage: `--dp [num_replicas]` and `--load-balance [heuristic]`

Example (data parallel = 2, tensor parallel = 2):
```
python -m  sglang.launch_server --model-path meta-llama/Llama-2-7b-chat-hf --tp 2 --dp 2 --port 30000 --load-balance shortest_queue
```
Note:
* This PR only include static launch of local data parallel replicas. (Elastic data parallel coming next)
* This PR only include dispatch heuristics of `round_robin` and `shortest_queue`.

## Refactor
This PR refactors some names/classes
```
sglang/srt/router -> sglang/srt/controller
sglang/srt/router/manager.py -> sglang/srt/controller/manager_single.py
sglang/srt/router/model_rpc.py -> sglang/srt/controller/tp_worker.py
```

Co-authored by: @xiezhq-hermann @hnyls2002

### CODE REVIEW

File: python/sglang/srt/server_args.py [hunk 1]
```diff
@@ -226,6 +230,23 @@ def add_cli_args(parser: argparse.ArgumentParser):
             help="Set API key of the server",
         )
 
+        # distributed arguments
+        parser.add_argument(
+            "--dp-size",
+            type=int,
+            default=ServerArgs.dp_size,
+            help="Number of tensor parallelism group for data parallelism.",
```
Comment (merrymercy) [inline python/sglang/srt/server_args.py]:
```suggestion
            help="Data parallelism size.",
```

File: python/sglang/srt/server_args.py [hunk 2]
```diff
@@ -226,6 +230,23 @@ def add_cli_args(parser: argparse.ArgumentParser):
...
+        )
+        parser.add_argument(
+            "--dispatch-method",
+            type=str,
+            default=ServerArgs.dispatch_method,
```
Comment (merrymercy) [inline python/sglang/srt/server_args.py]:
```suggestion
            default=ServerArgs.dispatch_method,
            help="Load balancing strategy for data parallelism.",
```

File: python/sglang/srt/utils.py [hunk 3]
```diff
@@ -138,30 +141,48 @@ def is_port_available(port):
...
-        ret_ports = [port]
-
-    ret_ports = list(set(x for x in ret_ports if is_port_available(x)))
-    cur_port = ret_ports[-1] + 1 if len(ret_ports) > 0 else 10000
-
-    while len(ret_ports) < 5 + tp_size:
-        if cur_port not in ret_ports and is_port_available(cur_port):
-            ret_ports.append(cur_port)
-        cur_port += 1
+    ret_ports = []
```
Comment (merrymercy) [inline python/sglang/srt/utils.py]: Use this new logic https://github.com/sgl-project/sglang/blob/55c16436273d4a42f7cfe342df5f10ad05a8d0fe/python/sglang/srt/utils.py#L141-L164

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: python/sglang/srt/managers/controller/manager_multi.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+        # Init communication
+        context = zmq.asyncio.Context()
+        self.recv_from_tokenizer = context.socket(zmq.PULL)
+        self.recv_from_tokenizer.bind(f"tcp://127.0.0.1:{port_args.router_port}")
+
+        # Init status
+        self.recv_reqs = []
+
+        # Start data parallel workers
+        # TODO: parallelize this
```
Comment (Ying1123) [inline python/sglang/srt/managers/controller/manager_multi.py]:
```suggestion
```

**Review by merrymercy — APPROVED** [review]

**Pull request merged**